javascript/packages/fury/lib/gen/collection.ts

```diff
@@ -0,0 +1,365 @@
+  /** Whether track elements ref. */
+  TRACKING_REF: 0b1,
+
+  /** Whether collection has null. */
+  HAS_NULL: 0b10,
+
+  /** Whether collection elements type is not declare type. */
+  NOT_DECL_ELEMENT_TYPE: 0b100,
+
+  /** Whether collection elements type different. */
+  NOT_SAME_TYPE: 0b1000,
+};
+
+class CollectionAnySerializer {
+  constructor(private fury: Fury) {
+
+  }
+
+  protected writeElementsHeader(arr: any) {
+    let flag = 0;
+    let isSame = true;
+    let serializer: Serializer | null = null;
+    let includeNone = false;
+
+    for (const item of arr) {
+      if ((item === undefined || item === null) && !includeNone) {
+        includeNone = true;
+      }
+      if (isSame) {
+        const current = this.fury.classResolver.getSerializerByData(item);
+        if (serializer !== null && current !== null && current !== serializer) {
+          isSame = false;
+        }
+        if (current !== null) {
+          serializer = current;
+        }
+      }
+    }
+
+    if (!isSame) {
+      flag |= CollectionFlags.NOT_SAME_TYPE;
+    }
+    if (includeNone) {
+      flag |= CollectionFlags.HAS_NULL;
+    }
+    if (serializer !== null && serializer.meta.needToWriteRef) {
+      flag |= CollectionFlags.TRACKING_REF;
+    }
+    this.fury.binaryWriter.uint8(flag);
+    if (isSame) {
+      if (serializer) {
+        this.fury.binaryWriter.int16(serializer.meta.type);
+      } else {
+        this.fury.binaryWriter.skip(2);
+      }
+    }
+    return {
+      serializer,
+      isSame,
+      flag,
+      includeNone,
+    };
+  }
+
+  write(value: any, size: number) {
+    const { serializer, isSame, includeNone } = this.writeElementsHeader(value);
+    this.fury.binaryWriter.varUInt32(size);
+    if (isSame) {
+      if (!includeNone) {
+        for (const item of value) {
+          serializer!.write(item);
+        }
+        return;
+      } else {
+        for (const item of value) {
+          if (item === null || item === undefined) {
+            this.fury.binaryWriter.uint8(RefFlags.NullFlag);
+            continue;
+          }
+          this.fury.binaryWriter.uint8(RefFlags.NotNullValueFlag);
+          serializer!.write(item);
+        }
+      }
+    } else {
+      if (includeNone) {
+        for (const item of value) {
+          const serializer = this.fury.classResolver.getSerializerByData(item);
+          if (item === null || item === undefined) {
+            this.fury.binaryWriter.uint8(RefFlags.NullFlag);
+            continue;
+          }
+          this.fury.binaryWriter.uint8(RefFlags.NotNullValueFlag);
+          this.fury.binaryWriter.uint16(serializer!.meta.type);
+          serializer!.write(item);
+        }
+      } else {
+        for (const item of value) {
+          const serializer = this.fury.classResolver.getSerializerByData(item);
+          this.fury.binaryWriter.uint16(serializer!.meta.type);
+          serializer!.write(item);
+        }
+      }
+    }
```

Comment:
Can something be streamlined here? A lot of repetition.

Reply:
Indeed, there is a lot of duplicate code; I'll think about how to simplify it.

Reply:
Whether to write the 2-byte type and the 1-byte refflag is determined by whether the element contains null or is entirely of the same type. I can't think of a better way to simplify this without avoiding the determination of 'isSame' and 'includeNone' on each iteration.

Reply:
```javascript
    for (const item of value) {
      if (includeNone) {
        if (item === null || item === undefined) {
          this.fury.binaryWriter.uint8(RefFlags.NullFlag);
          continue;
        }
        this.fury.binaryWriter.uint8(RefFlags.NotNullValueFlag);
      }

      let finalSerializer = serializer;
      if (!isSame) {
        finalSerializer = this.fury.classResolver.getSerializerByData(item);
        this.fury.binaryWriter.uint16(finalSerializer!.meta.type);
      }
      finalSerializer!.write(item);
    }
```

Can it become like this?

Reply:
LGTM. Although this introduces `isSame` and `includeNone` check overhead for every element, but compared to other memory writing overhead, I think the overhead acceptable